### PR TITLE
Basic Miniscript support in output descriptors

### PR DESCRIPTION
**Pull request by sipa**
This pull request introduces support for [miniscript](http://bitcoin.sipa.be/miniscript) in Bitcoin Core.

The bulk of the code is in the 3 commits that add the miniscript module, including conversion from/to CScript, converting to and parsing from its engineer-readable string notation, property analysis and ops limit/stack size limit that are necessary to assess the security of arbitrary scripts.

A number of tests are included, including tests against known scripts, and against randomly generated scripts.

The final commit integrates the miniscript module into descriptors. This is only rudimentary, as it is not yet integrated in the signing code. I'm including it here to give something accessible to play with, but if desirable I can move that to a later PR as well.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

Thanks for the great work on miniscript and the C++ implementation!

This version seems to deviate somewhat to the version in the [upstream repo](https://github.com/sipa/miniscript) which makes it unclear to me which of the issues I found during my review that has been addressed:
* [Avoid termination due to unhandled exception when parsing invalid arguments to after(...), older(...), thresh(...) and thresh_m(...)](https://github.com/sipa/miniscript/pull/3)
* [Avoid assertion failure on after(k) or older(k) where k < 1 or k >= 0x80000000UL](https://github.com/sipa/miniscript/pull/4)
* [Avoid running out of stack space by limiting recursion depth in Parse(...)](https://github.com/sipa/miniscript/pull/5)
* [Policies with too high nesting depth are not rejected: It is possible to create small inputs that cause extreme memory usage (in practice: OOM kill or std::bad_alloc)](https://github.com/sipa/miniscript/issues/7)

Could you clarify? :-)

File: src/script/miniscript.h [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+template<typename Key> using NodeRef = std::shared_ptr<const Node<Key>>;
+
+//! Construct a miniscript node as a shared_ptr.
+template<typename Key, typename... Args>
+NodeRef<Key> MakeNodeRef(Args&&... args) { return std::make_shared<const Node<Key>>(std::forward<Args>(args)...); }
+
+//! The different node types in miniscript.
+enum class NodeType {
+    FALSE,     //!< OP_0
+    TRUE,      //!< OP_1
```
Comment (MarcoFalke) [inline src/script/miniscript.h]: Looks like these already are macros on mac from `/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk/usr/include/mach/boolean.h`?

And also in windows: https://travis-ci.org/bitcoin/bitcoin/jobs/580527642#L3384

File: src/test/miniscript_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        std::copy(first, last, keyid.begin());
+        auto it = g_testdata->pkmap.find(keyid);
+        assert(it != g_testdata->pkmap.end());
+        key = it->second;
+        return true;
+    }
+};
+
+//! Singleton instance of KeyConverter.
+const KeyConverter CONVERTER;
```
Comment (MarcoFalke) [inline src/test/miniscript_tests.cpp]:
```
test/miniscript_tests.cpp:115:20: error: default initialization of an object of const type 'const (anonymous namespace)::KeyConverter' without a user-provided default constructor
```

https://travis-ci.org/bitcoin/bitcoin/jobs/580527646#L3225

File: src/script/miniscript.h [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+template<typename Key, typename Ctx>
+inline NodeRef<Key> Parse(Span<const char>& in, const Ctx& ctx) {
+    auto expr = Expr(in);
+    // Parse wrappers
+    for (int i = 0; i < expr.size(); ++i) {
+        if (expr[i] == ':') {
+            auto in2 = expr.subspan(i + 1);
+            auto sub = Parse<Key>(in2, ctx);
+            if (!sub || in2.size()) return {};
+            for (size_t j = i; j-- > 0; ) {
```
Comment (MarcoFalke) [inline src/script/miniscript.h]:
```
Running tests: miniscript_tests from test/miniscript_tests.cpp
Running 2 test cases...
Test cases order is shuffled using seed: 1580527639
Entering test module "Bitcoin Core Test Suite"
test/miniscript_tests.cpp(282): Entering test suite "miniscript_tests"
test/miniscript_tests.cpp(322): Entering test case "random_tests"
script/miniscript.h:599:33: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'size_t' (aka 'unsigned long')

Comment (sipa) [reply src/script/miniscript.h]: Thanks for investigating; fixed.

File: src/test/miniscript_tests.cpp (hunk 2 above)
Comment (sipa) [reply src/test/miniscript_tests.cpp]: Fixed pedantry by using `const KeyConverter CONVERTER{};` instead.

File: src/script/miniscript.h (hunk 1 above)
Comment (sipa) [reply src/script/miniscript.h]: Fixed by renaming to `JUST_0` and `JUST_1`.

**Comment by sipa** [comment]
@practicalswift Thanks for reminding me, I had forgotten about those (and wasn't really looking at the miniscript repo while working on this PR). Specifically:
* The overflow/underflow issue and the OOM issue don't apply here, as they're in the compiler code which I'm not PR'ing here.
* The unhandled exception I've fixed independently because `stoul` and friends are locale dependent (both here and upstream).
* The stack depth we need to fix still, I've commented on the issue.

Also, this PR should be perfectly in sync with the upstream repo, so if you want to PR fixes, they can go there preferably, and I'll incorporate them here.

File: src/script/miniscript.h [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+                int32_t sat_sum = 0;
+                std::vector<int32_t> diffs;
+                for (const auto& sub : subs) {
+                    dsat += sub->ss.dsat.value; // The type system requires "d" for thresh, so dsat must always be valid.
+                    if (sub->ss.sat.valid) diffs.push_back((int32_t)sub->ss.sat.value - (int32_t)sub->ss.dsat.value);
+                }
+                if (diffs.size() < k) return {{}, dsat};
+                std::sort(diffs.begin(), diffs.end());
+                for (size_t i = diffs.size() - k; i < diffs.size(); ++i) sat_sum += diffs[i];
+                return {sat_sum + dsat, dsat};
```
Comment (MarcoFalke) [inline src/script/miniscript.h]: Looks like this is still failing on https://bitcoinbuilds.org/index.php?job=e1623137-779c-47dd-8a8a-ebaee60766ec

```
Entering test module "Bitcoin Core Test Suite"
test/miniscript_tests.cpp(282): Entering test suite "miniscript_tests"
test/miniscript_tests.cpp(284): Entering test case "fixed_tests"
test/miniscript_tests.cpp(284): Leaving test case "fixed_tests"; testing time: 35249us
test/miniscript_tests.cpp(322): Entering test case "random_tests"
script/miniscript.h:508:33: runtime error: unsigned integer overflow: 4294967295 + 11 cannot be represented in type 'unsigned int'
    #0 0x561901402481 in miniscript::Node<CPubKey>::CalcStackSize() const /home/ubuntu/src/src/./script/miniscript.h:508:33

Comment (sipa) [reply src/script/miniscript.h]: Fixed by switching to a different algorithm.

**Comment by promag** [comment]
Concept ACK.

You could submit the refactors as separate pulls one by one. For instance, one for fc4691899d128cd4d8a8b60503803fc53d43eb9f which I think is acceptable on its own.

**Comment by practicalswift** [comment]
@sipa Would you mind cherry-picking in the `miniscript::FromScript(...)` from #17129 in to this PR? That would allow for closing #17129 which is entirely dependent on the merge of this one anyway :)

**Comment by practicalswift** [comment]
@fanquake Could you add "Waiting for author"? :)

**Comment by MarcoFalke** [comment]
"Needs rebase" implies "Waiting for author"

**Comment by achow101** [comment]
I've noticed that there are some Miniscript functions that may conflict in naming or functionality with existing output descriptors that may cause issues.

* `multi()` and `thresh_m()` are functionally the same. So the easy thing to do would be to just make `multi()` an alias for `thresh_m()`
* `sortedmulti()` currently has no Miniscript counterpart. But I think that is easily solved by having a `sorted_thresh_m()` which is semantically the same.
* `pkh()` and `pk_h()` are very similar, both in naming and function. `pkh()` is the same as `c:pk_h()` so I think we can just call `pkh()` an alias for `c:pk_h()`. Since `pkh()` and `pk_h()` are named differently, the only other issue is that they may confuse people.
* `pk()` in descriptors directly conflicts with `pk()` in Miniscript. `pk()` in descriptors is `c:pk()` in Miniscript. Because they are named exactly the same but have (slightly) different meanings, this can cause some issues.
  * One solution is to change the meaning of `pk()` depending on its location. If found at the top level or as the direct child of a `sh()` or `wsh()`, it would mean the current meaning today (to preserve compatibility). If found elsewhere, it would have the Miniscript meaning. But this solution is confusing as `pk()` has multiple meanings and it could be easy to mess up.
  * Alternatively we could deprecate and remove the usage of `pk()` and replace it with `c:pk()`. We could add `c:pk()` now (for 0.20) and have it mean the same thing as `pk()` now. Then once Miniscript is merged, remove `pk()`'s descriptor definition and just use the Miniscript definition. We could add warnings in some places as well to let people know that their usage of `pk()` is deprecated. The good thing about this is that `pk()` probably isn't currently being used so this probably won't effect a lot of people.

We will still need to preserve the current naming of functions as people may already be using them and changing these would break compatibility. At least most of these are straightforward aliases.

**Comment by kittiphan2405** [comment]
ขอบคุณสำหรับคำแนะนำครับ.. แต่คือทำทุกอย่างแล้วครับ....
ผมยังไม่เจอรูปแบบไหนที่ยืนยันตัวตน.. แล้วระบบไม่หลุดบ้างไม่มีดลย..
ผมพยายามจะเปิด.กระเป๋าหลายใบเพืรอเก็บทรัพ.. ตอนนี้ผมยังไม่เจออะไรเลย..
เสีย9้อมูลเปล่าๆ

กิตติพันธ์

ในวันที่ ส. 8 ก.พ. 2020 02:12 น. Andrew Chow <notifications@github.com>
เขียนว่า:

**Comment by Sjors** [comment]
Concept ACK, after many hours of @apoelstra explaining Miniscript and it's relation to Output Descriptors :-)

I like @achow101's suggestion:
>   * Alternatively we could deprecate and remove the usage of `pk()` and replace it with `c:pk()`. We could add `c:pk()` now (for 0.20) and have it mean the same thing as `pk()` now.

Would be great to get a rebase of this before 0.20 splits off, so we can deprecate other Output Descriptor stuff early on, if needed.

**Comment by sipa** [comment]
After a number of discussions, I think this highlights the possibility to decrease the gap between Miniscript and non-Miniscript descriptors.

My suggestion is making the following changes to Miniscript:
* Rename `thresh_m` to `multi`.
* Rename `pk` to `pk_k`.
* Add an alias `pk(A)=c:pk_k(A)` (just like `and_n(A,B)=andor(A,B,0)` for example).
* Add an alias `pkh(A)=c:pk_h(A)`.

That means that for users, there doesn't need to be a distinction between Miniscript or not, as `pk(A)` and `wsh(pk(A))` would look uniform. Meanwhile, implementations could choose what to deal with at what layer (e.g. top-level `pkh` could be dealt with as a special case, or through a sufficiently generic Miniscript library).

**Comment by practicalswift** [comment]
@sipa I would like to continue my robustness testing of this PR. Could you please rebase on `master` and perhaps also pull in my fix for the heap out-of-bounds read present in the current version of this PR: ["Avoid heap out-of-bounds read in Node::CalcOps (test case: OP_0 OP_2 OP_EQUAL) and assertion failure in ComputeType (test case: OP_0 OP_0 OP_EQUAL)"](https://github.com/sipa/miniscript/pull/18)? :)

**Comment by darosior** [comment]
Concept ACK

**Comment by michaelfolkson** [comment]
Scratching the surface on what would be needed to revitalize this PR. Obviously there is a mega rebase to do as it has been sitting here for a while but there have also been some changes since to the [C++ implementation](https://github.com/sipa/miniscript) of Miniscript. Presumably it would make sense to work on the Core rebase first and then the Miniscript updates?

**Comment by meshcollider** [comment]
(Just for the information of those wondering what's happening here, there is currently a few things being finished up in the miniscript repository before this PR is rebased, but it is a current WIP so expect something soon-ish!)

**Comment by Sjors** [comment]
It would be nice to have an up to date version here to compare with ElementsProject/libwally-core#310

**Comment by darosior** [comment]
The most up-to-date WIP is available here
https://github.com/darosior/bitcoin/pull/2 but for the purpose of reviewing other implementations i'd recommend checking against the C++ implem at sipa/miniscript or the Rust one at rust-bitcoin/rust-miniscript.
-------- Original Message --------

**Comment by fanquake** [comment]
@sipa did you want to close this now that #24147 is open?

**Comment by sipa** [comment]
Superseded by #24147.

**Pull request closed without merging**